core/src/alpenglow_consensus/voting_loop.rs

```diff
@@ -885,86 +943,97 @@ impl VotingLoop {
+            else {
+                warn!(
                 "Vote account {vote_account_pubkey} has no authorized voter for epoch {}.  Unable to vote",
                 bank.epoch()
             );
-            return GenerateVoteTxResult::Failed;
-        };
+                return GenerateVoteTxResult::Failed;
+            };
 
-        let authorized_voter_keypair = match authorized_voter_keypairs
-            .iter()
-            .find(|keypair| keypair.pubkey() == authorized_voter_pubkey)
-        {
-            None => {
-                warn!(
-                    "The authorized keypair {authorized_voter_pubkey} for vote account \
+            authorized_voter_keypair = match authorized_voter_keypairs
+                .iter()
+                .find(|keypair| keypair.pubkey() == authorized_voter_pubkey)
+            {
+                None => {
+                    warn!(
+                        "The authorized keypair {authorized_voter_pubkey} for vote account \
                      {vote_account_pubkey} is not available.  Unable to vote"
-                );
-                return GenerateVoteTxResult::NonVoting;
+                    );
+                    return GenerateVoteTxResult::NonVoting;
+                }
+                Some(authorized_voter_keypair) => authorized_voter_keypair.clone(),
+            };
+        }
+
+        let bls_keypair = match Self::get_bls_keypair(context, &authorized_voter_keypair) {
+            Ok(keypair) => keypair,
+            Err(e) => {
+                warn!("Failed to derive my own BLS keypair: {e:?}");
+                return GenerateVoteTxResult::FailedToDeriveBlsKeypair;
             }
-            Some(authorized_voter_keypair) => authorized_voter_keypair,
         };
-
-        let vote_ix =
-            vote.to_vote_instruction(vote_account_pubkey, authorized_voter_keypair.pubkey());
-        let vote_tx = Transaction::new_signed_with_payer(
-            &[vote_ix],
-            Some(&context.identity_keypair.pubkey()),
-            &[&context.identity_keypair, authorized_voter_keypair],
-            bank.last_blockhash(),
-        );
-
+        let my_bls_pubkey = bls_keypair.public.into();
+        let vote_serialized = bincode::serialize(&vote).unwrap();
+        let signature = authorized_voter_keypair.sign_message(&vote_serialized);
         if !context.has_new_vote_been_rooted {
-            context.voted_signatures.push(vote_tx.signatures[0]);
+            context.voted_signatures.push(signature);
             if context.voted_signatures.len() > MAX_VOTE_SIGNATURES {
                 context.voted_signatures.remove(0);
             }
         } else {
             context.voted_signatures.clear();
         }
 
-        GenerateVoteTxResult::Tx(vote_tx)
+        let Some(my_rank) = Self::get_my_rank(context, bank, &my_bls_pubkey) else {
+            return GenerateVoteTxResult::NoRankFound;
```

Comment:
hmm this error seems like this would fail silently if say somebody misconfigured or never swapped over to BLS since the callers never log:

https://github.com/anza-xyz/alpenglow/blob/6e3110dd18d5b23c3f81fc364d867706d1cd6d2f/core/src/alpenglow_consensus/voting_loop.rs#L594

https://github.com/anza-xyz/alpenglow/blob/6e3110dd18d5b23c3f81fc364d867706d1cd6d2f/core/src/alpenglow_consensus/voting_loop.rs#L441-L447

https://github.com/anza-xyz/alpenglow/blob/6e3110dd18d5b23c3f81fc364d867706d1cd6d2f/core/src/alpenglow_consensus/voting_loop.rs#L360-L366

Could we add an error log here/in the caller `send_vote()` as well?

Reply:
Outside send_vote the error reason is lost, since only bool is returned. I added log inside send_vote, do we need duplicate log in the caller as well?